lib/WGwrangler/IPmanager.pm

```diff
@@ -0,0 +1,268 @@
+        },
+        ...
+    }
+
+=cut
+sub ipv4_build_database($self, $ref_ip_list) {
+    my $ip_database = $self->{ip_database};
+    my $ip_meta_info = $self->{ip_meta_info};
+    for my $target_name (keys %{$ref_ip_list}) {
+        my $ip_target_database = {};
+        my $n_ips_for_target = 0;
+        for my $networks (@{$ref_ip_list->{$target_name}}) {
+            my ($network, $subnet_size) = @{$networks};
+
+            if ($subnet_size < MAX_SUBNET_SIZE) {
+                warn "Subnet size ($subnet_size) is larger than MAX_SUBNET_SIZE, skipping $target_name";
+                next;
+            }
+            if (defined $network) {
+                my (undef, $n_ips) = _generate_ipv4_list($network, $subnet_size, $ip_target_database);
+                $n_ips_for_target += $n_ips;
+            }
+        }
+        $ip_database->{$target_name} = $ip_target_database;
+        $ip_meta_info->{$target_name}{n_ips} = $n_ips_for_target;
+    }
+}
+
+=head3 ipv4_acquire($ipv4, $target_name)
+
+Tries to acquire one ip from a target.
+
+B<Parameters>
+
+=over 1
+
+=item
+
+C<$ipv4> An ip address as string e.g I<10.0.30.56>
+
+=item
+
+C<$target_name> A target name
+
+=back
+
+B<Returns>
+
+Returns 1 on success, undef on failure
+
+=cut
+sub ipv4_acquire($self, $ipv4, $target_name) {
+    chomp $ipv4;
+    if (exists $self->{ip_database}{$target_name} && $self->{ip_database}{$target_name}{$ipv4}) {
+        delete $self->{ip_database}{$target_name}{$ipv4};
+        $self->{ip_meta_info}{$target_name}{n_ips}--;
+        return 1;
+    }
+    else {
+        return undef;
+    }
+}
+
+=head3 ipv4_release($ipv4, $target_name)
+
+Tries to acquire one ip from a target.
+
+B<Caveat:> Due tue they way this utility is implemented, this function basically adds C<$ipv4> to list
+of available ips, regardless of the fact that it may not fit in the sourcing subnet.
+
+B<Parameters>
+
+=over 1
+
+=item
+
+C<$ipv4> An ip address as string e.g I<10.0.30.56>
+
+=item
+
+C<$target_name> A target name
+
+=back
+
+B<Returns>
+
+Returns 1 on success, undef on failure.
+
+=cut
+sub ipv4_release($self, $ipv4, $target_name) {
+    if (exists $self->{ip_database}{$target_name}) {
+        $self->{ip_database}{$target_name}{$ipv4} = 1;
+        $self->{ip_meta_info}{$target_name}{n_ips}++;
+        return 1;
+    }
+    else {
+        return undef;
+    }
+}
+
+sub ipv4_suggest($self, $target_name, $n = 1) {
+    if (exists $self->{ip_database}{$target_name} && $self->{ip_meta_info}{$target_name}{n_ips} > 0) {
+        $n = $self->{ip_meta_info}{$target_name}{n_ips} - 1 if $n >= $self->{ip_meta_info}{$target_name}{n_ips};
+        my @free_ips = keys %{$self->{ip_database}{$target_name}};
+        return @free_ips[0 .. $n];
+    }
+    else {
+        return ();
+    }
+}
+
+
+sub _generate_ipv4_list($network_id, $subnet_size, $ref_ip_database) {
+    # thanks to https://www.perl.com/article/creating-ip-address-tools-from-scratch/
+
+    my $n_ips = 0;
```

Comment:
there are some pretty powerful perl modules for managing ip addresses around ... including ipv6 support

Reply:
Indeed, migrated to Net::IP - however, I had to drop ip suggestions for now...